state/consumer.go

```diff
@@ -47,7 +53,16 @@ func (k *ConsumersCollection) Add(consumer Consumer) error {
 	if !utils.Empty(consumer.Username) {
 		searchBy = append(searchBy, *consumer.Username)
 	}
-	_, err := getConsumer(txn, searchBy...)
+	_, err := getConsumer(txn, []string{"Username", "id"}, searchBy...)
+	if err == nil {
+		return fmt.Errorf("inserting consumer %v: %w", consumer.Console(), ErrAlreadyExists)
+	} else if err != ErrNotFound {
+		return err
+	}
```

Comment:
The way I understand the algorithm implemented in this function is:
1. Try to find consumer by ID.
2. If the above yielded not result, if username not empty, try to find the consumer by username.
3. If the above yielded no result, try to find the consumer by custom ID.
4. If any of the above yielded anything other than ErrNotFound, fail insertion.

Is it right? If so, the implementation appears somewhat complicated for reasons I don't understand - it first performs a multi-index lookup searching for both username and ID in both indices (on the condition of non-empty username), and then takes a different approach by doing a single-index lookup on custom ID.

Would a more streamlined alternative (pseudocode) work as well? Like:

```go
_, err := getConsumer(txn, []string{"id"}, *consumer.ID)
// return error if err == nil or err != ErrNotFound
_, err := getConsumer(txn, []string{"Username"}, *consumer.Username)
// return error if err == nil or err != ErrNotFound
_, err := getConsumer(txn, []string{"CustomID"}, *consumer.CustomID)
// return error if err == nil or err != ErrNotFound
```

Furthermore, the "find a potential collision on either ID or Username or CustomID" should, given its complexity, imo be wrapped in a function - "checkCustomerAlreadyExists" or something like that.

Reply:
Also, should we test this (preventing a duplicate CustomID from being inserted)?

Reply:
That's the "refactoring across the board to search only a single index" approach. It's fine, but we'd then reasonably do the same for other resources. I'm not aware of others that'd actually need it, so I chose to preserve the status quo.